File: src/rust/src/simulator.rs
//! Simulation controller

use crate::robot::{Robot, RobotState};
use crate::environment::Environment;
use crate::pathfinding::AStarPlanner;

/// Simulation configuration
pub struct SimulationConfig {
    pub max_steps: usize,
    pub enable_slam: bool,
    pub tick_rate: f64,
}

impl Default for SimulationConfig {
    fn default() -> Self {
        Self {
            max_steps: 10000,
            enable_slam: false,
            tick_rate: 0.1,
        }
    }
}

/// Main simulator
pub struct Simulator {
    pub robot: Robot,
    pub environment: Environment,
    pub config: SimulationConfig,
    pub steps: usize,
}

impl Simulator {
    /// Create new simulator
    pub fn new(robot: Robot, environment: Environment, config: SimulationConfig) -> Self {
        Self {
            robot,
            environment,
            config,
            steps: 0,
        }
    }

    /// Execute one simulation step
    pub fn step(&mut self) -> bool {
        self.steps += 1;
        self.environment.step(self.config.tick_rate);

        // Simple simulation logic
        match self.robot.state {
            RobotState::Idle => {
                self.robot.state = RobotState::Cleaning;
            }
            RobotState::Cleaning => {
                // Check if should return to dock
                if self.robot.should_return_to_dock() {
                    self.robot.state = RobotState::ReturningToDock;
                }
            }
            RobotState::Charging => {
                if self.robot.charge(10.0) {
                    self.robot.state = RobotState.Cleaning;
                }
            }
            _ => {}
        }

        // Check max steps
        if self.steps >= self.config.max_steps {
            return false;
        }

        true
    }

    /// Run complete simulation
    pub fn run(&mut self) -> SimulationResults {
        log::info!("Starting simulation");

        while self.step() {
            // Simulation loop
        }

        log::info!("Simulation complete after {} steps", self.steps);

        SimulationResults {
            steps: self.steps,
            cleaning_coverage: self.environment.get_cleaning_percentage(),
            total_distance: self.robot.stats.total_distance,
            battery_cycles: self.robot.stats.battery_cycles,
            success: self.robot.state != RobotState::Error,
        }
    }
}

/// Simulation results
#[derive(Debug, Clone)]
pub struct SimulationResults {
    pub steps: usize,
    pub cleaning_coverage: f64,
    pub total_distance: f64,
    pub battery_cycles: usize,
    pub success: bool,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::Position;

    #[test]
    fn test_simulator_creation() {
        let robot = Robot::new(Position::new(15.0, 15.0));
        let env = Environment::create_empty_room(30, 30);
        let config = SimulationConfig::default();

        let sim = Simulator::new(robot, env, config);

        assert_eq!(sim.steps, 0);
    }

    #[test]
    fn test_simulator_step() {
        let robot = Robot::new(Position::new(15.0, 15.0));
        let env = Environment::create_empty_room(30, 30);
        let config = SimulationConfig::default();

        let mut sim = Simulator::new(robot, env, config);

        let should_continue = sim.step();

        assert!(should_continue);
        assert_eq!(sim.steps, 1);
    }
}

File: src/julia/RobotVacuum/src/simulator.jl
"""
Simulation controller for robot vacuum cleaner.
"""

"""
    SimulationConfig

Configuration for simulation run.

# Fields
- `room_type::String`: Type of room to simulate
- `cleaning_mode::CleaningMode`: Cleaning mode to use
- `max_steps::Int`: Maximum simulation steps
- `enable_slam::Bool`: Enable SLAM
- `random_seed::Union{Int,Nothing}`: Random seed for reproducibility
"""
struct SimulationConfig
    room_type::String
    cleaning_mode::CleaningMode
    max_steps::Int
    enable_slam::Bool
    random_seed::Union{Int,Nothing}
end

function SimulationConfig(;
    room_type::String="furnished",
    cleaning_mode::CleaningMode=Auto,
    max_steps::Int=10000,
    enable_slam::Bool=true,
    random_seed::Union{Int,Nothing}=nothing
)
    SimulationConfig(room_type, cleaning_mode, max_steps, enable_slam, random_seed)
end

"""
    Simulator

Main simulation controller.

# Fields
- `robot::Robot`: Robot instance
- `environment::Environment`: Environment instance
- `config::SimulationConfig`: Configuration
- `steps::Int`: Current step count
- `slam::Union{SLAM,Nothing}`: SLAM system (if enabled)
"""
mutable struct Simulator
    robot::Robot
    environment::Environment
    config::SimulationConfig
    steps::Int
    slam::Union{SLAM,Nothing}
end

function Simulator(robot::Robot, environment::Environment, config::SimulationConfig)
    slam = config.enable_slam ? SLAM(environment.width, environment.height) : nothing

    Simulator(robot, environment, config, 0, slam)
end

"""
    step!(sim::Simulator) -> Bool

Execute one simulation step. Returns true if simulation should continue.
"""
function step!(sim::Simulator)::Bool
    sim.steps += 1
    step!(sim.environment)

    # Simple simulation logic
    if sim.robot.state == Idle
        sim.robot.state = Cleaning
    elseif sim.robot.state == Cleaning
        if should_return_to_dock(sim.robot)
            sim.robot.state = ReturningToDock
        end
    elseif sim.robot.state == Charging
        if charge!(sim.robot)
            sim.robot.state = Cleaning
        end
    end

    # Check max steps
    sim.steps >= sim.config.max_steps ? false : true
end

"""
    run_simulation(config::SimulationConfig) -> Dict

Run complete simulation and return results.
"""
function run_simulation(config::SimulationConfig)::Dict{String,Any}
    @info "Starting simulation"

    if !isnothing(config.random_seed)
        Random.seed!(config.random_seed)
    end

    # Create environment based on room type
    env = if config.room_type == "empty"
        create_empty_room(50, 50)
    elseif config.room_type == "furnished"
        create_room_with_furniture(50, 50, 5)
    else
        create_empty_room(50, 50)
    end

    # Find starting position (center)
    start_pos = Position(env.width / 2, env.height / 2)

    # Create robot
    robot = Robot(start_pos)
    robot.mode = config.cleaning_mode

    # Create simulator
    sim = Simulator(robot, env, config)

    # Run simulation
    while step!(sim)
        # Simulation loop
    end

    @info "Simulation complete after $(sim.steps) steps"

    # Return results
    Dict{String,Any}(
        "steps" => sim.steps,
        "cleaning_coverage" => get_cleaning_percentage(sim.environment),
        "total_distance" => sim.robot.stats.total_distance,
        "battery_cycles" => sim.robot.stats.battery_cycles,
        "success" => sim.robot.state != Error
    )
end
